x-pack/plugins/siem/server/lib/detection_engine/signals/filter_events_with_list.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { get } from 'lodash/fp';
+import { Logger } from 'src/core/server';
+
+import { ListItemArraySchema } from '../../../../../lists/common/schemas/response';
+import { type } from '../../../../../lists/common/schemas/common';
+import { ListClient } from '../../../../../lists/server';
+import { SignalSearchResponse, SearchTypes } from './types';
+import { RuleAlertParams } from '../types';
+
+interface FilterEventsAgainstList {
+  listClient: ListClient;
+  exceptionsList: RuleAlertParams['exceptions_list'];
+  logger: Logger;
+  eventSearchResult: SignalSearchResponse;
+}
+
+export const filterEventsAgainstList = async ({
+  listClient,
+  exceptionsList,
+  logger,
+  eventSearchResult,
+}: FilterEventsAgainstList): Promise<SignalSearchResponse> => {
+  try {
+    if (exceptionsList == null || exceptionsList.length === 0) {
+      return eventSearchResult;
+    }
+
+    // narrow unioned type to be single
+    const isStringableType = (val: SearchTypes) =>
+      ['string', 'number', 'boolean'].includes(typeof val);
+    // grab the signals with values found in the given exception lists.
+    const filteredHitsPromises = exceptionsList.map(async (exceptionItem) => {
+      if (exceptionItem.values == null || exceptionItem.values.length === 0) {
+        throw new Error('Malformed exception list provided');
+      }
+      if (!type.is(exceptionItem.values[0].name)) {
+        throw new Error(`Unsupported list type used, please use one of ${type.keys}`);
+      }
+      if (!exceptionItem.values[0].id) {
+        throw new Error(`Missing list id for exception on field ${exceptionItem.field}`);
+      }
+      // acquire the list values we are checking for.
+      const valuesOfGivenType = eventSearchResult.hits.hits.reduce((acc, searchResultItem) => {
+        const valueField = get(exceptionItem.field, searchResultItem._source);
+        if (valueField != null && isStringableType(valueField) && !acc.has(valueField.toString())) {
+          acc.add(valueField.toString());
+        }
+        return acc;
+      }, new Set<string>());
+
+      const listSignals: ListItemArraySchema = await listClient.getListItemByValues({
```

Comment:
nit: You don't need the extra cast of `const listSignals: ListItemArraySchema =`, you can just do: `const listSignals =` reduces some noise.